genie-agent/src/main/java/com/netflix/genie/agent/cli/ExecCommand.java

```diff
@@ -45,26 +49,52 @@
     private final ExecCommandArguments execCommandArguments;
     private final JobExecutionStateMachine stateMachine;
     private final KillService killService;
-    private final List<sun.misc.Signal> signalsToIntercept = Collections.unmodifiableList(Arrays.asList(
-        new sun.misc.Signal("INT"),
-        new sun.misc.Signal("TERM")
-    ));
+    private final ThreadFactory threadFactory;
+    private final ShutdownProperties shutdownProperties;
+
+    private boolean isRunning;
+    private final ReentrantLock isRunningLock = new ReentrantLock();
+    private final Condition isRunningCondition = this.isRunningLock.newCondition();
 
     ExecCommand(
         final ExecCommandArguments execCommandArguments,
         final JobExecutionStateMachine stateMachine,
-        final KillService killService
+        final KillService killService,
+        final AgentProperties agentProperties
+    ) {
+        this(
+            execCommandArguments,
+            stateMachine,
+            killService,
+            agentProperties,
+            Thread::new
+        );
+    }
+
+    @VisibleForTesting
+    ExecCommand(
+        final ExecCommandArguments execCommandArguments,
+        final JobExecutionStateMachine stateMachine,
+        final KillService killService,
+        final AgentProperties agentProperties,
+        final ThreadFactory threadFactory // For testing shutdown hooks
     ) {
         this.execCommandArguments = execCommandArguments;
         this.stateMachine = stateMachine;
         this.killService = killService;
+        this.shutdownProperties = agentProperties.getShutdown();
+        this.threadFactory = threadFactory;
     }
 
     @Override
     public ExitCode run() {
-        for (final sun.misc.Signal s : signalsToIntercept) {
-            sun.misc.Signal.handle(s, signal -> handleTerminationSignal());
-        }
+
+        // Lock-free since the only other thread accessing this has not been registered yet
+        this.isRunning = true;
+
+        // Before execution starts, add shutdown hooks
+        Runtime.getRuntime().addShutdownHook(this.threadFactory.newThread(this::waitForCleanShutdown));
```

Comment:
General comment: if we can ever get the Agent to java 9+ we can leverage improvements in [Process API](https://www.baeldung.com/java-9-process-api) that may help with things beyond just the shutdown hook.